### ISSUE TITLE
Security concerns over longevity of tokens

### ISSUE THREAD

**Issue by ningyuansg** [source-academy/cadet#53]
In today's meeting, the SoC systems engineer raised concerns over the security of the long-lived (52 weeks) refresh token. He recommends a validity period more similar to IVLE (8 hours).

**Comment by indocomsoft** [comment]
This was Evan's idea, not sure why he wants me to set it that long.

I was also not sure about Evan's rationale but I feel like it should be the refresh token that requires revocation on logout (instead of access token). The access token will expire in 1 hour anyway. I think this measure will help mitigate the security concern. Do tell me what y'all think.

However, I feel that 8 hours is way too short. Imagine the hassle if you have to login when you access the website the next day. One other solution we could do is to have a `remember_me` flag which gives longer refresh token life (say, 7 days) which can be exchanged for a new refresh token by the client right before expiry using a new endpoint on the backend.

To summarise, I propose the access token validity period be 1 hour, refresh token be between 8-24 hours, with an option to have a longer refresh token validity period of 7 days. Refresh token is to be sent to the server upon logout to be added to the revocation list. There should be an endpoint to exchange soon-expiring refresh token for a new one.

This should combine the security of short-lived access token and reducing exposure of long-lived refresh token while still allowing us to revoke refresh token and reducing the number of times we hit the database (revocation check done only on refresh token).

Let me know what you think on this issue @ningyuansg @remo5000 @tuesmiddt @sreycodes

**Comment by ningyuansg** [comment]
How about the same 1 hour access token, 52 weeks refresh token, but we only allow one valid refresh token at any one time?

**Comment by indocomsoft** [comment]
@ningyuansg That wouldn't work if the user is using multiple devices which seems like a very arbitrary restriction.

Tbh I don't really see the issue with the 52 weeks refresh token as long as we ensure communication is via HTTPS and front-end protect against XSS (and CSRF if you're using cookies but I doubt it.) I am actually more concerned with being able to exchange refresh token with a newer one without re-sending credentials.

That, and ensuring revocation can be done on the refresh token.

**Comment by indocomsoft** [comment]
Bump @ningyuansg @remo5000 @tuesmiddt @sreycodes

**Comment by indocomsoft** [comment]
Ok so for now, I decided to refactor according to this:

1 hour access token, 7 days refresh token. Only refresh token is revocable (stored in db). Client needs to submit refresh token to be revoked to the server through the `logout` endpoint.

@ningyuansg @remo5000 I will update the swagger as well.

**Comment by ningyuansg** [comment]
Sorry, I missed your last post on this issue. Sounds good to me.
